File: backend/src/main/java/org/icpclive/events/EventsLoader.java
package org.icpclive.events;

import org.icpclive.Config;
import org.icpclive.events.PCMS.PCMSEventsLoader;
import org.icpclive.events.PCMS.ioi.IOIPCMSEventsLoader;
import org.icpclive.events.WF.json.WFEventsLoader;
import org.icpclive.events.codeforces.CFEventsLoader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Properties;

public abstract class EventsLoader implements Runnable {
    private static final Logger log = LoggerFactory.getLogger(EventsLoader.class);

    private static EventsLoader instance;

    protected double emulationSpeed;
    protected long emulationStartTime;

    public static synchronized EventsLoader getInstance() {
        if (instance == null) {
            try {
                Properties properties = Config.loadProperties("events");
                String standingsType = properties.getProperty("standings.type");
                if ("WF".equals(standingsType)) {
                    instance = new WFEventsLoader(false);
                } else if ("WFRegionals".equals(standingsType)) {
                    instance = new WFEventsLoader(true);
                } else if ("PCMS".equals(standingsType)) {
                    instance = new PCMSEventsLoader();
                } else if ("CF".equals(standingsType)) {
                    instance = new CFEventsLoader();
                } else if ("IOIPCMS".equals(standingsType)) {
                    instance = new IOIPCMSEventsLoader();
                }
            } catch (IOException e) {
                log.error("error", e);
            }
        }
        return instance;
    }


    public abstract ContestInfo getContestData();

    public double getEmulationSpeed() {
        return emulationSpeed;
    }
}

File: backend/src/main/kotlin/org/icpclive/events/EventsLoader.kt
package org.icpclive.events

import org.icpclive.Config.loadProperties
import org.icpclive.events.PCMS.PCMSEventsLoader
import org.icpclive.events.WF.json.WFEventsLoader
import org.icpclive.events.codeforces.CFEventsLoader

abstract class EventsLoader : Runnable {
    var emulationSpeed = 0.0
        protected set
    protected var emulationStartTime: Long = 0
    abstract val contestData: ContestInfo?

    companion object {
        val instance by lazy {
            val properties = loadProperties("events")
            val standingsType = properties.getProperty("standings.type")
            when (standingsType) {
                "WF" -> WFEventsLoader(false)
                "WFRegionals" -> WFEventsLoader(true)
                "PCMS" -> PCMSEventsLoader()
                "CF" -> CFEventsLoader()
                else -> throw IllegalArgumentException("Unknown standings.type $standingsType")
            }
        }
    }
}